### ISSUE TITLE
Support for `postgres_fdw` in GreptimeDB to enable integration with PostgreSQL

### ISSUE THREAD

**Issue by frin1** [GreptimeTeam/greptimedb#5077]
### What problem does the new feature solve?

The current PostgreSQL protocol implementation in GreptimeDB does not fully support Foreign Data Wrappers (FDWs), particularly the `postgres_fdw` extension. This limitation makes it challenging to integrate GreptimeDB with existing PostgreSQL databases. By adding this support, it would be easier to use PostgreSQL as a metadata store while querying and interacting with GreptimeDB for time-series data. This would allow seamless integration of GreptimeDB in PostgreSQL-based systems and enhance its adoption in existing PostgreSQL workflows.

### What does the feature do?

This feature would enable `postgres_fdw` to interact with GreptimeDB by:
- Supporting required PostgreSQL protocol features to make `postgres_fdw` functional.
- Allowing users to define foreign tables in PostgreSQL that link directly to tables in GreptimeDB.

The functionality would allow users to query GreptimeDB tables (e.g. `numbers`) directly from PostgreSQL.

### Implementation challenges

1. **Protocol Compliance**: Ensuring that GreptimeDB's PostgreSQL protocol implementation meets the requirements of `postgres_fdw`.
2. **Query Translation**: Mapping queries issued via `postgres_fdw` into GreptimeDB's query execution engine. Differences in SQL dialects or functions might need careful handling.
3. **Performance Optimization**: Avoiding inefficiencies caused by translating queries or transferring large volumes of time-series data between systems.
4. **Testing and Compatibility**: Ensuring that the feature works with common PostgreSQL versions and setups where `postgres_fdw` is used.

**Comment by killme2008** [comment]
@frin1 It's interesting! I don't known `postgres_fdw` before.

@sunng87 Is it posibble to make `pgwire` supporting `postgres_fdw`?

**Comment by sunng87** [comment]
Thank you for the information. I just tried foreign data wrapper by myself. At wire protocol level we are fully compatible. It's mostly unsupported SQL statements:

- some postgres specific transaction statements and we can bypass them with #5081 
- cursor: the foreign data wrapper uses a cursor mechanism to fetch data from remote server https://www.postgresql.org/docs/current/plpgsql-cursors.html

The statements ran by fdw are like:

- `DECLARE c1 CURSOR FOR\nSELECT id FROM public.tree`
- `FETCH 100 FROM c1`
- `CLOSE c1`

Currently we don't have cursor in greptimedb so this is the major road blocker. I tried to find if there is an option to disable this behaviour but with no luck https://www.postgresql.org/docs/current/postgres-fdw.html

**Comment by frin1** [comment]
@sunng87 Thank you for the response and for testing the foreign data wrapper functionality.

I see you have created a new pull request to add more transaction-related statements for the PostgreSQL interface. Do you think `postgres_fdw` will work after this is implemented, or is cursor support still a blocker? Is there any plan to implement all the features that `postgres_fdw` requires? I believe this would be beneficial for GreptimeDB, as it would enable seamless integration between a PostgreSQL database for metadata and GreptimeDB for time-series data.

**Comment by sunng87** [comment]
@frin1 Yes. I agree being able to use GreptimeDB as a foreign data provider can be very useful for those who use postgres to manage their IoT metadata.

I'm prototyping the cursor support for greptimedb now. Hopefully we can have a working demo with postgre_fdw in a few days.

**Comment by frin1** [comment]
> @frin1 Yes. I agree being able to use GreptimeDB as a foreign data provider can be very useful for those who use postgres to manage their IoT metadata.
> 
> I'm prototyping the cursor support for greptimedb now. Hopefully we can have a working demo with postgre_fdw in a few days.

Great! Looking forward to testing it out. Thanks for working on this!

**Comment by sunng87** [comment]
Implemented in #5094 and #5081 
I will add docs next week.

**Comment by frin1** [comment]
@sunng87 Awesome! Thanks for implementing this so quickly.
